### PR TITLE
ENH: pass values through shlex_quote, fixes #3625

### LINKED ISSUE #3625: Shlex_quote() values instead of ""-ing placeholders in run-procedure
Noting down an issue @kyleam brought to attention in https://github.com/datalad/datalad/pull/3624#discussion_r317657539:

Within 
https://github.com/datalad/datalad/blob/c9abcb0b3c4e91bb3df1721c8bcd2d7d60371461/datalad/interface/run_procedure.py#L182-L193

placeholders should be stripped of the quotes, and values should be passed through shlex_quote().

### PR DESCRIPTION
**Pull request by adswa**
In this PR I have

- removed quoted placeholders from the template in ``_guess_exec`` in ``run_procedure.py`` and
- instead shlex_quoted() the values supplied in ``__call__``, (if we are not on Windows)

<details>
<summary> However, locally, this crashes two tests: </summary>

```
======================================================================
FAIL: datalad.core.local.tests.test_create.test_create_withprocedure
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/adina/env/handbook/lib/python3.7/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/adina/repos/datalad/datalad/tests/utils.py", line 607, in newfunc
    return t(*(arg + (filename,)), **kw)
  File "/home/adina/repos/datalad/datalad/core/local/tests/test_create.py", line 406, in test_create_withprocedure
    eq_(ds.config['datalad.metadata.nativetype'], ('xmp', 'datacite'))
AssertionError: '"xmp" "datacite"' != ('xmp', 'datacite')

======================================================================
FAIL: datalad.interface.tests.test_run_procedure.test_configs
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/adina/env/handbook/lib/python3.7/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/adina/repos/datalad/datalad/tests/utils.py", line 895, in newfunc
    return func(*args, **kwargs)
  File "/home/adina/repos/datalad/datalad/tests/utils.py", line 434, in newfunc
    return t(*(arg + (d,)), **kw)
  File "/home/adina/repos/datalad/datalad/interface/tests/test_run_procedure.py", line 228, in test_configs
    ok_file_has_content(op.join(ds.path, 'fromproc.txt'), 'some_arg\n')
  File "/home/adina/repos/datalad/datalad/tests/utils.py", line 417, in ok_file_has_content
    assert_equal(content, file_content, **kwargs)
AssertionError: 'some_arg\n' != '"some_arg"\n'
- some_arg
+ "some_arg"
? +        +
```
</details>

because multiple arguments are now returned as a single string. I lack the judgment to decide whether this is indeed problematic, and at which point I can do adjustments. 

Can someone give me a pointer?

(I put the commits on top of those in https://github.com/datalad/datalad/pull/3624 to prevent merge conflicts, let me know if that's okay)

### CODE REVIEW

File: datalad/interface/run_procedure.py [hunk 1]
```diff
@@ -445,9 +449,10 @@ def __call__(
                              "Missing 'execute' permissions?" % procedure_file)
 
         cmd = ex['template'].format(
-            script=procedure_file,
-            ds=ds.path if ds else '',
-            args=u' '.join(u'"{}"'.format(a) for a in args) if args else '')
+            script=procedure_file if on_windows else shlex_quote(procedure_file),
+            ds='' if not ds else (ds.path if on_windows else shlex_quote(ds.path)),
+            args=shlex_quote(u' '.join(u'"{}"'.format(a) for a in args) if args else '') if not on_windows
+                    else u' '.join(u'"{}"'.format(a) for a in args) if args else '')
```
[the remark this answers is not in the document]
Comment (adswa) [reply datalad/interface/run_procedure.py]: oh, thx!

**Comment by mih** [comment]
FTR: The test might be fixed like this:

```diff
diff --git a/datalad/interface/tests/test_run_procedure.py b/datalad/interface/tests/test_run_procedure.py
index d9a76919e..4a17b7765 100644
--- a/datalad/interface/tests/test_run_procedure.py
+++ b/datalad/interface/tests/test_run_procedure.py
@@ -238,7 +238,7 @@ def test_configs(path):
     # for run:
     ds.config.add(
         'datalad.procedures.datalad_test_proc.call-format',
-        'python "{script}" "{ds}" {{mysub}} {args}',
+        'python {script} {ds} {{mysub}} {args}',
         where='dataset'
     )
     ds.config.add(
@@ -258,7 +258,7 @@ def test_configs(path):
     # config on dataset level:
     ds.config.add(
         'datalad.procedures.datalad_test_proc.call-format',
-        'python "{script}" "{ds}" local {args}',
+        'python {script} {ds} local {args}',
         where='local'
     )
     ds.unlock("fromproc.txt")
```

I could not test it though, because the current PR is against 0.11, rebase wasn't straightforward and with all the extensions my dev environment cannot easily handle this. Sorry.

**Comment by adswa** [comment]
@mih this makes it work for me, I've pushed those changes. Thanks!

File: datalad/utils.py [hunk 2]
```diff
@@ -2279,5 +2280,13 @@ def get_suggestions_msg(values, known, sep="\n        "):
...
+    """
+    shlex_quote() a value if the command is not run on a Windows
+    machine.
+    """
+
+    return val if on_windows else shlex_quote(val)
+
 
-lgr.log(5, "Done importing datalad.utils")
+lgr.log(5, "Done importing datalad.utils")
```
Comment (mih) [inline datalad/utils.py]: I personally prefer line endings all the way to the last line, but we are not agreement within the project already.

File: datalad/interface/run_procedure.py [hunk 3]
```diff
@@ -15,7 +15,9 @@
 
 from glob import iglob
 from argparse import REMAINDER
+from six.moves import shlex_quote
```
Comment (mih) [inline datalad/interface/run_procedure.py]: This is no longer needed with `maybe_shlex_quote`

File: datalad/interface/run_procedure.py [hunk 4]
```diff
@@ -33,6 +35,7 @@
 from datalad.distribution.dataset import datasetmethod
 from datalad.support.exceptions import InsufficientArgumentsError
 from datalad.support.exceptions import NoDatasetArgumentFound
+from datalad.utils import on_windows, maybe_shlex_quote
```
Comment (mih) [inline datalad/interface/run_procedure.py]: It is not used consistently in the code base (yet), but we now do (for any import with more than a single item)

```
from datalad.utils import (
    on_windows,
    maybe_shlex_quot,
)
```

i.e. one per line, each line (incl last one has a trailing comma) -- this minimizes the diffs and make them quicker to grasp.

That being said `on_windows` is no longer needed, see below.

File: datalad/interface/run_procedure.py [hunk 5]
```diff
@@ -181,15 +184,16 @@ def _guess_exec(script_file):
...
-                'template': u'"{script}" "{ds}" {args}',
+                'template': u'{script} {ds} {args}',
                 'state': state}
     elif script_file.endswith('.sh'):
         return {'type': u'bash_script',
-                'template': u'bash "{script}" "{ds}" {args}',
+                'template': u'bash {script} {ds} {args}',
                 'state': state}
     elif script_file.endswith('.py'):
+        ex = sys.executable if on_windows else shlex_quote(sys.executable)
```
Comment (mih) [inline datalad/interface/run_procedure.py]: Can also be `maybe_shlex_quote(sys.executable)` now

Comment (adswa) [reply datalad/interface/run_procedure.py]: good point, thx.

File: datalad/interface/run_procedure.py (hunk 4 above)
Comment (adswa) [reply datalad/interface/run_procedure.py]: Ah yes, I missed this. I removed all superfluous imports now.

File: datalad/utils.py (hunk 2 above)
Comment (adswa) [reply datalad/utils.py]: done.

**Comment by mih** [comment]
Tests did not run (apt issue). Restarted.

File: datalad/interface/tests/test_run_procedure.py [hunk 6]
```diff
@@ -258,7 +258,7 @@ def test_configs(path):
     # config on dataset level:
     ds.config.add(
         'datalad.procedures.datalad_test_proc.call-format',
-        'python "{script}" "{ds}" local {args}',
+        'python {script} {ds} local {args}',
```
Comment (mih) [inline datalad/interface/tests/test_run_procedure.py]: Here and above we should also use `sys.executable` not `python`.

Comment (adswa) [reply datalad/interface/tests/test_run_procedure.py]: oh right. Thanks!

**Comment by adswa** [comment]
Hi @kyleam, I have tried to write a regression test, but that ended up being a test with args that contain spaces -- which is a test that I think cannot fail currently (although there currently is no test that explicitly gives args with spaces, so I pushed it anyway). I wasn't able to come up with any test that can trigger a failure in the case of quoted placeholders... If you have an idea how to write a test for this, you can let me know or add one yourself, if you like.

**Comment by mih** [comment]
Thx @adswa  and @kyleam !

**Pull request merged**